augur/export_v2.py

```diff
@@ -827,7 +827,10 @@ def set_description(data_json, cmd_line_description_file):
 
 def parse_node_data_and_metadata(T, node_data_files, metadata_file):
     node_data = read_node_data(node_data_files) # node_data_files is an array of multiple files (or a single file)
-    metadata, _ = read_metadata(metadata_file) # metadata={} if file isn't read / doeesn't exist
+    if metadata_file:
```

Comment:
This test should explicitly check for whether `metadata is not None`. For example, when a user doesn't specify the metadata file argument in an `augur export v2` call, the value of `metadata` is `None` and indicates that the user didn't try to provide an argument. If the user passes an empty string (unintentionally or not), we should treat that as an error (passing the string to the `read_metadata` function and letting it raise a `FileNotFoundError`).

We also need to add a try/except block around the `parse_node_data_and_metadata` call in `run_v2` to catch the file not file error and other exceptions, print the error messages to stderr, and exit with a non-zero code.

Reply:
Changed check to `if metadata_file is not None`, preserving the legacy behavior of using `{}` for the metadata in that case.

I'm not sure we need to catch any exceptions in export_v2, unless it's important to exit in a certain way instead of letting the exception bubble up. Especially since that module seems to be ripe for refactoring into a few classes in a future PR.

Reply:
In this case, we need to exit in a certain way because the `run_v2` function is executed at the command line interface. If the user passes an empty string for metadata right now with the following command,

```
augur export v2 --tree results/tree.nwk --metadata "" --node-data results/branch_lengths.json --output analysis.json
```

 they will get a messy exception traceback like this:

```
Traceback (most recent call last):
  File "../pulls/augur/bin/augur", line 24, in <module>
    exit( main() )
  File "../pulls/augur/augur/__main__.py", line 10, in main
    return augur.run( argv[1:] )
  File "../pulls/augur/augur/__init__.py", line 74, in run
    return args.__command__.run(args)
  File "../pulls/augur/augur/export.py", line 22, in run
    return run_v2(args)
  File "../pulls/augur/augur/export_v2.py", line 881, in run_v2
    node_data, node_attrs, node_data_names, metadata_names = parse_node_data_and_metadata(T, args.node_data, args.metadata)
  File "../pulls/augur/augur/export_v2.py", line 831, in parse_node_data_and_metadata
    metadata, _ = read_metadata(metadata_file)
  File "../pulls/augur/augur/utils.py", line 74, in read_metadata
    return MetadataFile(fname, query).read()
  File "../pulls/augur/augur/util_support/metadata_file.py", line 18, in __init__
    self.key_type = self.find_key_type()
  File "../pulls/augur/augur/util_support/metadata_file.py", line 60, in find_key_type
    if "strain" not in self.columns and "name" not in self.columns:
  File "../pulls/augur/augur/util_support/metadata_file.py", line 57, in columns
    return list(self.parse_file().columns)
  File "../pulls/augur/augur/util_support/metadata_file.py", line 81, in parse_file
    skipinitialspace=True,
  File "/Users/jlhudd/miniconda3/envs/nextstrain/lib/python3.6/site-packages/pandas/io/parsers.py", line 676, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/Users/jlhudd/miniconda3/envs/nextstrain/lib/python3.6/site-packages/pandas/io/parsers.py", line 448, in _read
    parser = TextFileReader(fp_or_buf, **kwds)
  File "/Users/jlhudd/miniconda3/envs/nextstrain/lib/python3.6/site-packages/pandas/io/parsers.py", line 880, in __init__
    self._make_engine(self.engine)
  File "/Users/jlhudd/miniconda3/envs/nextstrain/lib/python3.6/site-packages/pandas/io/parsers.py", line 1114, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/Users/jlhudd/miniconda3/envs/nextstrain/lib/python3.6/site-packages/pandas/io/parsers.py", line 1891, in __init__
    self._reader = parsers.TextReader(src, **kwds)
  File "pandas/_libs/parsers.pyx", line 374, in pandas._libs.parsers.TextReader.__cinit__
  File "pandas/_libs/parsers.pyx", line 674, in pandas._libs.parsers.TextReader._setup_parser_source
FileNotFoundError: [Errno 2] File  does not exist: ''
```

A good example of how this is currently handled in export is [the function to read in the description markdown](https://github.com/nextstrain/augur/blob/4182fc5bc3bff647fba60a4ad6805cb2d786ad59/augur/export_v2.py#L821-L826).